src/core/index.ts

```diff
@@ -284,10 +291,36 @@ export class ZarrArray {
       return out;
     }
 
-    for (const proj of indexer.iter()) {
-      await this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes);
+    const concurrencyLimit = (
+      options === undefined
+      || options === null
+      || options.concurrencyLimit === undefined
+      || options.concurrencyLimit === null
+    ) ? 10 : options.concurrencyLimit;
+
+    const progressCallback = (
+      options === undefined
+      || options === null
+      || options.progressCallback === undefined
+      || options.progressCallback === null
+    ) ? null : options.progressCallback;
+
+    const limit = pLimit(concurrencyLimit);
+    const input = Array.from(
+      indexer.iter(),
+      proj => limit(() => this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes))
+    );
+
+    if (progressCallback) {
```

Comment:
I would personally just call it on every iteration instead of given some delay. This will keep firing indefinitely if an error is thrown I think.

Perhaps we can call the callback in the function which is `limit()`ed? 
Something like
```typescript
if (progressCallback) {
  const projections = Array.from(indexer.iter());
  const numChunks = allProjections.length;
  let progress = 0;
  for(const proj of indexer.iter()) {
    limit( () => {
      await this.chunkGetitem(proj.chunkCoords, ...etc)
      progress += 1;
      // I think we can just pass an object and let the customer decide whether they want 0..1 or something else :)
      progressCallback({progress: progress, total: numChunks})
      }
  }
} else {
const input = Array.from(
      indexer.iter(),
      proj => limit(() => this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes))
    );
}
```

I'm sure it can be done cleaner, but I hope you get the idea

Reply:
in the above example, I guess we should also call the progress callback before the first iteration with `{progress: 0, total: numChunks}`

Reply:
I see. I used the `setInterval` since I'm not sure if `await`-ing within the `limit()`-ed function will have the desired asynchronous behavior. Although, I agree, there should be better error handling here. 

Maybe there is a way to provide a callback to `limit`. I'll have a look. 

Returning a response object is smart.